doc/source/workflows/events.rst

```diff
@@ -99,6 +100,111 @@ After the workflow finishes checkpointing the event, the event listener will be
+---------------
+
+An HTTP event is a specific workflow event received from an HTTP endpoint enabled
+by Ray serve. The implementation of the HTTP endpoint is named `HTTPEventProvider`. Its
+corresponding event listener is `HTTPListener`.
+
+Receive an event by HTTP
+--------------------------
+
+`HTTPListener` is a customized `EventListener` that subscribes to an `HTTPEventProvider`
+with the unique combination of `workflow_id` and `event_key`. An HTTP request from
+an external event submitter needs to specify `workflow_id` as part of the request URL and
+a JSON formatted document consisting of `event_key` and `event_payload`.
+
+The external submitter learns the values of `workflow_id` and `event_key` through
+the application context, which is out of the scope in this discussion.
+
+.. code-block:: python
+
+    import ray
+    from ray import workflow
+    from ray.workflow.http_event_provider import HTTPListener
+
+    # Create a task waiting for an http event with a JSON message.
+    # The JSON message is expected to have an event_key field
+    # and an event_payload field.
+    event_task = workflow.wait_for_event(HTTPListener, event_key='my_event_key')
+
+    key, messsage = workflow.run(event_task, workflow_id='workflow_receive_event_by_http')
+
+workflow_id as part of the request URL
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, `HTTPEventProvider` is deployed as a Ray serve endpoint at
+`http://hostname:port/event/send_event/<workflow_id>`. If it is deployed locally,
+the endpoint may be reached at `http://127.0.0.1:8000/event/send_event/<workflow_id>`.
+Note that `workflow_id` must be included in the web request to match the active
+`wait_for_event` task.
+
+If there are more than one active `wait_for_event` tasks within the same
+`workflow_id`, `HTTPEventProvider` will forward to the task subscribing to
+the matching `event_key`.
+
+Expected event content in JSON
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, `HTTPEventProvider` expects an arriving message to be a parsable JSON
+document with two fields, `event_key` and `event_payload`, as shown in the example
+below. The `event_key` field must match the `event_key` argument of the `workflow.wait_for_event`
+task. The value of the `event_payload` field is returned, along with the value of
+the `event_key` field as the output of the task.
+
+.. code-block:: json
+
+    {"event_key": "my_event_key", "event_payload": "my_event_message"}
+
+Example event submitter client
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An external event submitter needs to have:
+
+* the HTTP endpoint address (e.g. `http://127.0.0.1:8000/event/send_event/`)
+* the `workflow_id` (e.g. "workflow_receive_event_by_http")
+* a valid JSON formatted message
+
+The code snippet below retries submission by checking the HTTP response code 404,
+which is returned by `HTTPEventProvider` if the `event_key` was not expected or
+the `workflow_id` did not exist, HTTP response code 200 is returned if the event
```

Comment:
We need to explain why it'll return 200.

Reply:
The paragraph now explains why.